src/util/firmware.ts

```diff
@@ -44,27 +47,53 @@ class FirmwareManager {
       status: this.currentStatus,
       message: this.message,
       error: this.errorSeen ? this.message : undefined,
+      percentage: this.percentage,
     };
   }
 
   private updateStatus(
     status: InstallStatus,
     message: string,
     isError = false,
+    percentage?: number,
   ) {
     this.currentStatus = status;
     this.message = message;
     this.errorSeen = isError;
+    if (percentage !== undefined) {
+      this.percentage = percentage;
+    }
+  }
+
+  private getLatestPercentage(log: string): number {
+    const lines = log.split('\n');
+    const percentageRegex = /(\d{1,3})%/;
+    // Start from the end to find the first line with a percentage
+    for (let i = lines.length - 1; i >= 0; i--) {
+      const match = lines[i].match(percentageRegex);
+      if (match) {
+        return parseInt(match[1], 10);
+      }
+    }
+    return 0;
   }
 
   private runSpawn(cmd: string) {
     this.updateStatus(InstallStatus.Installing, 'Installation in progress...');
 
     const child = spawn(cmd, { shell: true });
+    let output = '';
 
     child.stdout.setEncoding('utf8');
     child.stdout.on('data', data => {
-      this.updateStatus(InstallStatus.Installing, data.toString());
+      output += data.toString();
```

Comment:
Please correct me if I am missing something here:
we can also do: output = data.toString() right ?
as every line has percentage value
this way we do not have to loop through whole output and just parse last line

Reply:
if we are concerned that not every line has percentage, then we can probably store it in a var like "lastKnownPercent" value ?

Reply:
`output = data.toString()`
No , that would output only the streaming portion received like "...."

Reply:
`We are still unaware of progress of Unzip command right ?`
Im ok with that, it's just a few seconds

Reply:
> if we are concerned that not every line has percentage, then we can probably store it in a var like "lastKnownPercent" value ?
so this is still a possibility right ?

Reply:
not sure I understand this:
`if we are concerned that not every line has percentage, then we can probably store it in a var like "lastKnownPercent" value ?`

Reply:
I prefer not to store any value because we would need to "clean it"

Reply:
so the regex will give me the last percentage found... so if we are on 27%... and for the next 10 lines there is no percentage we still send 27